python/ray/data/_internal/planner/exchange/shuffle_impl.py

```diff
@@ -0,0 +1,88 @@
+import math
+from typing import List, Optional, Tuple, Union
+
+import numpy as np
+
+from ray.data._internal.delegating_block_builder import DelegatingBlockBuilder
+from ray.data._internal.planner.exchange.interfaces import ExchangeInterface
+from ray.data.block import Block, BlockAccessor, BlockExecStats, BlockMetadata
+
+
+class ShuffleImpl(ExchangeInterface):
```

Comment:
This is copied from https://github.com/ray-project/ray/blob/master/python/ray/data/_internal/shuffle_and_partition.py#L12 .